File: ruby/lib/Bill.rb
class Bill
	attr_accessor :user
	def initialize(user=nil)
		@grocery_costs = 0.0
		@other_costs = 0.0
		@user= user
	end
	def add_item(item, category, quantity, price)
		@other_costs += Float(quantity * price) if category != "groceries"
		@grocery_costs += Float(quantity * price) if category == "groceries"
	end

	def total_cost
		Float(@other_costs + @grocery_costs)
	end
	def non_grocery_cost
		@other_costs
	end

	def net_payable_amount
		if @user != nil
			disc_amt = Float(@other_costs * @user.discount()/100)
			total = total_cost() - disc_amt
			total = Float(total - 5*(Integer(total)/100)) if total >100.0
			return total
		end
		total_cost()
	end
end

File: java/src/org/sumitbisht/lib/Bill.java
package org.sumitbisht.lib;

public class Bill 
{
	
	private double groceryCosts =0.0d;
	private double otherCosts = 0.0d;
	User user = null;
	
	public void setBillUser(User user)
	{
		this.user = user;
	}
	
	public void addItem( String category, int quantity, double price)
	{
		if(category.equals("groceries"))
		{
			groceryCosts += quantity * price;
		}
		else
		{
			otherCosts += quantity * price;
		}
	}
	
	public double getCosts()
	{
		return groceryCosts + otherCosts;
	}
	
	public double getOtherCosts()
	{
		return otherCosts;
	}
	
	public double netPayableAmount()
	{
		if(user!=null)
		{
			double discounts = otherCosts * user.discount()/100;
			double totalCost = getCosts() - discounts;
			totalCost = totalCost - (int)(totalCost/100)*5;
			return totalCost;
		}
		return getCosts();
	}

}
